lib/src/modules/task.cc

```diff
@@ -128,11 +136,163 @@ static TaskCommand getTaskCommand(const fs::path &task_executable)
+    fs::create_directory(cache_dir);
+    fs::last_write_time(cache_dir, time(nullptr));
+    fs::permissions(cache_dir, TASK_FILE_RW_PERMS);
+    return cache_dir;
+}
+
+// Computes the sha256 of the file denoted by path. Assumes that
+// the file designated by "path" exists.
+static std::string calculateSha256(const std::string& path) {
+    const std::streamsize CHUNK_SIZE = 2 << 14;
+    char buffer[CHUNK_SIZE];
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+
+    auto md = EVP_sha256();
+    auto mdctx = EVP_MD_CTX_create();
+    EVP_DigestInit_ex(mdctx, md, nullptr);
+    boost::nowide::ifstream ifs(path);
+    while (ifs.read(buffer, CHUNK_SIZE)) {
+      EVP_DigestUpdate(mdctx, buffer, CHUNK_SIZE);
+    }
+    if (!ifs.eof()) {
+      throw Module::ProcessingError(lth_loc::format("Error while reading {1}", path));
+    }
+    EVP_DigestUpdate(mdctx, buffer, ifs.gcount());
+    EVP_DigestFinal_ex(mdctx, md_value, &md_len);
+    EVP_MD_CTX_destroy(mdctx);
+
+    char md_value_ascii[2*EVP_MAX_MD_SIZE+1] = {'\0'};
+    for (unsigned int i = 0; i < md_len; ++i) {
+      snprintf(md_value_ascii+2*i, 3, "%02x", md_value[i]);
+    }
+    return std::string(md_value_ascii);
+}
+
+// Callback for libcurl. First param. is the chunk of data that's been downloaded, next
+// two params are the size of a single data element and the # of data elements
+// respectively, while the last param. is the stream that the data will be written
+// to. See https://curl.haxx.se/libcurl/c/libcurl-tutorial.html for more details.
+// Note that the maximum size of the data in "ptr" is 16K.
+static size_t writeData(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+    return fwrite(ptr, size, nmemb, stream);
+}
+
+// wrapper to curl_easy_escape that frees the returned char* pointer
+// while updating the url
+static void updateUrl(std::string& url, CURL* curl, const std::string& to_add) {
+    char* to_add_escp = curl_easy_escape(curl, to_add.c_str(), 0);
+    url += std::string(to_add_escp);
+    curl_free(to_add_escp);
+}
+
+// Creates the URL for libcurl to use when making the donwload.
+// TODO: Include the master uri too!
+static std::string createUrl(CURL* curl, const lth_jc::JsonContainer& uri) {
+    std::string url = uri.get<std::string>("path");
+    auto params = uri.getWithDefault<lth_jc::JsonContainer>("params", lth_jc::JsonContainer());
+    if (params.empty()) {
+      return url;
+    }
+    url += "?";
+    for (auto& key : params.keys()) {
+      updateUrl(url, curl, key);
+      url += "=";
+      updateUrl(url, curl, params.get<std::string>(key));
+      url += "&";
+    }
+    return url;
+}
+
+// This method downloads the file at the specified url into the provided path
+// If something goes wrong during the download, a PXP-error should be returned.
+static void downloadTaskFile(const fs::path& path, const lth_jc::JsonContainer& uri) {
+    auto curl = curl_easy_init();
```

Comment:
Would using https://github.com/puppetlabs/leatherman/blob/master/curl/inc/leatherman/curl/client.hpp simplify this?